light/beacon/binary_merkle.go

```diff
@@ -0,0 +1,555 @@
+			hasher.Write(proofHash[:])
+		} else {
+			hasher.Write(proofHash[:])
+			hasher.Write(value[:])
+		}
+		hasher.Sum(value[:0])
+		index /= 2
+		proofIndex++
+	}
+	if proofIndex < len(proof) {
+		return common.Hash{}, false
+	}
+	return common.Hash(value), true
+}
+
+// ProofFormat defines the shape of a partial proof and allows traversing a subset of a tree
+type ProofFormat interface {
+	children() (left, right ProofFormat) // either both or neither should be nil
+}
+
+// ProofReader allows traversing and reading a tree structure or a subset of it.
+// Note: the hash of each traversed node is always requested. If the internal
+// hash is not available then subtrees are always traversed (first left, then right).
+// If internal hash is available then subtrees are only traversed if needed by the writer.
+type ProofReader interface {
+	children() (left, right ProofReader) // subtrees accessible if not nil
+	readNode() (MerkleValue, bool)       // hash should be available if children are nil (leaf node), optional otherwise (internal node)
+}
+
+// ProofWriter allow collecting data for a partial proof while a subset of a tree is traversed.
+type ProofWriter interface {
+	children() (left, right ProofWriter) // all non-nil subtrees are traversed
+	writeNode(MerkleValue)               // called for every traversed tree node (both leaf and internal)
+}
+
+// TraverseProof traverses a reader and a writer defined on the same tree
+// simultaneously, copies data from the reader to the writer (if writer is not nil)
+// and returns the root hash. At least the shape defined by the writer is traversed;
+// subtrees not required by the writer are only traversed (with writer == nil)
+// if the hash of the internal tree node is not provided by the reader.
+func TraverseProof(reader ProofReader, writer ProofWriter) (common.Hash, bool) {
+	var (
+		wl ProofWriter
+		wr ProofWriter
+	)
+	if writer != nil {
+		wl, wr = writer.children()
+	}
+	node, nodeAvailable := reader.readNode()
+	if nodeAvailable && wl == nil {
+		if writer != nil {
+			writer.writeNode(node)
+		}
+		return common.Hash(node), true
+	}
+	rl, rr := reader.children()
+	if rl == nil {
+		return common.Hash{}, false
+	}
+	lhash, ok := TraverseProof(rl, wl)
+	if !ok {
+		return common.Hash{}, false
+	}
+	rhash, ok := TraverseProof(rr, wr)
+	if !ok {
+		return common.Hash{}, false
+	}
+	if !nodeAvailable {
+		hasher := sha256.New()
+		hasher.Write(lhash[:])
+		hasher.Write(rhash[:])
+		hasher.Sum(node[:0])
+	}
+	if writer != nil {
+		writer.writeNode(node)
+	}
+	return common.Hash(node), true
+}
+
+// MultiProof stores a partial Merkle tree proof
+type MultiProof struct {
+	Format ProofFormat
+	Values MerkleValues
+}
+
+// ParseSSZMultiProof creates a MultiProof from a standard serialized format
+// (2 byte subtree sizes, then 32 byte leaf values and internal sibling hashes
+// in depth-first, left-to-right order).
+func ParseSSZMultiProof(proof []byte) (MultiProof, error) {
```

Comment:
Can you elaborate the format of passed in multiProof?
Looks like it has
(1) a single byte prefix, maybe version?
(2) 2 bytes slice, maybe indices?
(3) 32 bytes slice, maybe leaves+internal nodes?

Reply:
I added a detailed description.